File: sdk/python/fractal_lba_client.py
#!/usr/bin/env python3
"""
Fractal LBA + Kakeya FT Stack Python SDK (Phase 3 WP7)

Official Python client for submitting Proof-of-Computation Summaries to the
Fractal LBA backend with built-in signature support, retry logic, and
error handling.

Usage:
    from fractal_lba_client import FractalLBAClient

    client = FractalLBAClient(
        base_url="https://api.fractal-lba.example.com",
        tenant_id="tenant-001",
        signing_key="your-hmac-key",
        signing_alg="hmac"
    )

    response = client.submit_pcs(pcs_dict)
"""

import hashlib
import hmac
import json
import time
from typing import Dict, Any, Optional, List
from decimal import Decimal, ROUND_HALF_UP
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry


class FractalLBAError(Exception):
    """Base exception for Fractal LBA client errors"""
    pass


class SignatureError(FractalLBAError):
    """Raised when signature generation or verification fails"""
    pass


class ValidationError(FractalLBAError):
    """Raised when PCS validation fails"""
    pass


class APIError(FractalLBAError):
    """Raised when API returns an error"""
    def __init__(self, message: str, status_code: int, response: Optional[Dict] = None):
        super().__init__(message)
        self.status_code = status_code
        self.response = response


class FractalLBAClient:
    """
    Python SDK client for Fractal LBA API (Phase 3 WP7)

    Features:
    - Automatic PCS signing (HMAC-SHA256)
    - Retry with exponential backoff + jitter
    - Multi-tenant support
    - Request validation
    - Response handling
    """

    def __init__(
        self,
        base_url: str,
        tenant_id: Optional[str] = None,
        signing_key: Optional[str] = None,
        signing_alg: str = "none",
        timeout: int = 30,
        max_retries: int = 3,
        retry_backoff_factor: float = 1.0
    ):
        """
        Initialize Fractal LBA client

        Args:
            base_url: Base URL of the API (e.g., https://api.fractal-lba.example.com)
            tenant_id: Tenant ID for multi-tenant mode (Phase 3)
            signing_key: HMAC signing key (if alg=hmac)
            signing_alg: Signature algorithm ("none", "hmac", "ed25519")
            timeout: Request timeout in seconds
            max_retries: Maximum number of retry attempts
            retry_backoff_factor: Backoff factor for retries
        """
        self.base_url = base_url.rstrip('/')
        self.tenant_id = tenant_id
        self.signing_key = signing_key
        self.signing_alg = signing_alg
        self.timeout = timeout

        # Create session with retry logic
        self.session = requests.Session()
        retry_strategy = Retry(
            total=max_retries,
            backoff_factor=retry_backoff_factor,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["POST", "GET"]
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

        # Set default headers
        self.session.headers.update({
            "Content-Type": "application/json",
            "User-Agent": "fractal-lba-python-sdk/0.3.0"
        })

        if self.tenant_id:
            self.session.headers.update({"X-Tenant-Id": self.tenant_id})

    def submit_pcs(self, pcs: Dict[str, Any]) -> Dict[str, Any]:
        """
        Submit a Proof-of-Computation Summary

        Args:
            pcs: PCS dictionary with all required fields

        Returns:
            VerifyResult dict with acceptance status

        Raises:
            ValidationError: If PCS is invalid
            SignatureError: If signature generation fails
            APIError: If API returns error
        """
        # Validate PCS
        self._validate_pcs(pcs)

        # Sign PCS if signing is enabled
        if self.signing_alg != "none":
            pcs = self._sign_pcs(pcs)

        # Submit to API
        url = f"{self.base_url}/v1/pcs/submit"

        try:
            response = self.session.post(
                url,
                json=pcs,
                timeout=self.timeout
            )

            # Handle different response codes
            if response.status_code == 200:
                return response.json()  # Accepted
            elif response.status_code == 202:
                return response.json()  # Escalated
            elif response.status_code == 401:
                raise APIError(
                    "Signature verification failed",
                    status_code=401,
                    response=response.json() if response.text else None
                )
            elif response.status_code == 429:
                retry_after = response.headers.get("Retry-After", "10")
                raise APIError(
                    f"Rate limit exceeded. Retry after {retry_after}s",
                    status_code=429,
                    response=response.json() if response.text else None
                )
            else:
                raise APIError(
                    f"API error: {response.status_code}",
                    status_code=response.status_code,
                    response=response.json() if response.text else None
                )

        except requests.exceptions.RequestException as e:
            raise APIError(f"Request failed: {e}", status_code=0)

    def _validate_pcs(self, pcs: Dict[str, Any]) -> None:
        """Validate PCS structure and required fields"""
        required_fields = [
            "pcs_id", "schema", "version", "shard_id", "epoch", "attempt",
            "sent_at", "seed", "scales", "N_j", "coh_star", "v_star",
            "D_hat", "r", "regime", "budget", "merkle_root", "ft"
        ]

        for field in required_fields:
            if field not in pcs:
                raise ValidationError(f"Missing required field: {field}")

        # Validate schema
        if pcs["schema"] != "fractal-lba-kakeya":
            raise ValidationError(f"Invalid schema: {pcs['schema']}")

        # Validate bounds
        if not (0 <= pcs["coh_star"] <= 1.05):
            raise ValidationError(f"coh_star out of bounds: {pcs['coh_star']}")

        if not (0 <= pcs["r"] <= 1):
            raise ValidationError(f"r out of bounds: {pcs['r']}")

        if not (0 <= pcs["budget"] <= 1):
            raise ValidationError(f"budget out of bounds: {pcs['budget']}")

        if pcs["regime"] not in ["sticky", "mixed", "non_sticky"]:
            raise ValidationError(f"Invalid regime: {pcs['regime']}")

    def _sign_pcs(self, pcs: Dict[str, Any]) -> Dict[str, Any]:
        """
        Sign PCS using configured algorithm

        Returns copy of PCS with 'sig' field populated
        """
        pcs_copy = pcs.copy()

        if self.signing_alg == "hmac":
            if not self.signing_key:
                raise SignatureError("HMAC key not configured")

            # Generate signature using Phase 1 canonicalization
            signature = self._sign_hmac(pcs_copy)
            pcs_copy["sig"] = signature

        elif self.signing_alg == "ed25519":
            raise SignatureError("Ed25519 signing not yet implemented in SDK")

        else:
            raise SignatureError(f"Unknown signing algorithm: {self.signing_alg}")

        return pcs_copy

    def _sign_hmac(self, pcs: Dict[str, Any]) -> str:
        """
        Generate HMAC-SHA256 signature (Phase 1 compatibility)

        Uses 8-field signature subset with 9-decimal rounding
        """
        # Extract signature subset (Phase 1 spec)
        subset = {
            "budget": self._round9(pcs["budget"]),
            "coh_star": self._round9(pcs["coh_star"]),
            "D_hat": self._round9(pcs["D_hat"]),
            "epoch": pcs["epoch"],
            "merkle_root": pcs["merkle_root"],
            "pcs_id": pcs["pcs_id"],
            "r": self._round9(pcs["r"]),
            "shard_id": pcs["shard_id"]
        }

        # Canonical JSON (sorted keys, no spaces)
        canonical_json = json.dumps(subset, sort_keys=True, separators=(',', ':'))

        # SHA-256 digest
        digest = hashlib.sha256(canonical_json.encode('utf-8')).digest()

        # HMAC-SHA256
        sig_bytes = hmac.new(
            self.signing_key.encode('utf-8'),
            digest,
            hashlib.sha256
        ).digest()

        # Base64 encode
        import base64
        return base64.b64encode(sig_bytes).decode('utf-8')

    def _round9(self, x: float) -> float:
        """Round float to 9 decimal places (Phase 1 spec)"""
        return float(Decimal(str(x)).quantize(Decimal("0.000000001"), rounding=ROUND_HALF_UP))

    def health_check(self) -> bool:
        """
        Check if API is healthy

        Returns:
            True if healthy, False otherwise
        """
        try:
            url = f"{self.base_url}/health"
            response = self.session.get(url, timeout=5)
            return response.status_code == 200
        except:
            return False

    def close(self):
        """Close session"""
        self.session.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


# Example usage
if __name__ == "__main__":
    # Example PCS
    example_pcs = {
        "pcs_id": "a7b3c8d9e0f1234567890abcdef12345",
        "schema": "fractal-lba-kakeya",
        "version": "0.1",
        "shard_id": "shard-001",
        "epoch": 12345,
        "attempt": 1,
        "sent_at": "2025-01-15T10:30:00Z",
        "seed": 42,
        "scales": [2, 4, 8, 16, 32],
        "N_j": {"2": 3, "4": 5, "8": 9, "16": 17, "32": 31},
        "coh_star": 0.85,
        "v_star": [0.12, 0.98, -0.05],
        "D_hat": 1.35,
        "r": 0.65,
        "regime": "sticky",
        "budget": 0.55,
        "merkle_root": "a1b2c3d4e5f6789012345678901234567890abcdefabcdef1234567890abcd",
        "sig": "",  # Will be populated by client
        "ft": {
            "outbox_seq": 123,
            "degraded": False,
            "fallbacks": [],
            "clock_skew_ms": 0
        }
    }

    # Create client
    client = FractalLBAClient(
        base_url="http://localhost:8080",
        tenant_id="tenant-001",
        signing_key="testsecret",
        signing_alg="hmac"
    )

    # Check health
    if client.health_check():
        print("✓ API is healthy")
    else:
        print("✗ API is not healthy")
        exit(1)

    # Submit PCS
    try:
        result = client.submit_pcs(example_pcs)
        print(f"✓ PCS submitted successfully")
        print(f"  Accepted: {result['accepted']}")
        print(f"  Escalated: {result['escalated']}")
        if 'recomputed_D_hat' in result:
            print(f"  Recomputed D̂: {result['recomputed_D_hat']:.4f}")
    except Exception as e:
        print(f"✗ Failed to submit PCS: {e}")
        exit(1)

File: sdk/ts/fractal-lba-client.ts
// TypeScript SDK for Fractal LBA + Kakeya FT Stack API (Phase 4 WP5)
import * as crypto from 'crypto';

/**
 * Proof-of-Computation Summary structure
 */
export interface PCS {
  pcs_id: string;
  schema: string;
  version: string;
  shard_id: string;
  epoch: number;
  attempt: number;
  sent_at: string;
  seed: number;
  scales: number[];
  N_j: Record<string, number>;
  coh_star: number;
  v_star: number[];
  D_hat: number;
  r: number;
  regime: string;
  budget: number;
  merkle_root: string;
  sig?: string;
  ft: FaultToleranceInfo;
}

/**
 * Fault tolerance metadata
 */
export interface FaultToleranceInfo {
  outbox_seq: number;
  degraded: boolean;
  fallbacks: string[];
  clock_skew_ms: number;
}

/**
 * Verification result from backend
 */
export interface VerifyResult {
  accepted: boolean;
  recomputed_D_hat?: number;
  recomputed_budget?: number;
  reason?: string;
  escalated: boolean;
}

/**
 * Client configuration options
 */
export interface ClientOptions {
  baseURL: string;
  tenantID?: string;
  signingKey?: string;
  signingAlg?: 'hmac' | 'none';
  timeout?: number;
  maxRetries?: number;
}

/**
 * Custom error types
 */
export class FractalLBAError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'FractalLBAError';
  }
}

export class SignatureError extends FractalLBAError {
  constructor(message: string) {
    super(message);
    this.name = 'SignatureError';
  }
}

export class ValidationError extends FractalLBAError {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

export class APIError extends FractalLBAError {
  constructor(
    message: string,
    public statusCode: number,
    public responseBody?: string
  ) {
    super(message);
    this.name = 'APIError';
  }
}

/**
 * Fractal LBA API Client (Phase 4 WP5)
 *
 * Implements Phase 1 canonical signing with automatic HMAC-SHA256 signing,
 * multi-tenant support, retry logic, and validation.
 *
 * @example
 * const client = new FractalLBAClient({
 *   baseURL: 'https://api.example.com',
 *   tenantID: 'tenant1',
 *   signingKey: 'supersecret',
 *   signingAlg: 'hmac'
 * });
 *
 * const result = await client.submitPCS(pcs);
 * console.log('Accepted:', result.accepted);
 */
export class FractalLBAClient {
  private baseURL: string;
  private tenantID: string;
  private signingKey: string;
  private signingAlg: 'hmac' | 'none';
  private timeout: number;
  private maxRetries: number;

  constructor(options: ClientOptions) {
    this.baseURL = options.baseURL;
    this.tenantID = options.tenantID || '';
    this.signingKey = options.signingKey || '';
    this.signingAlg = options.signingAlg || 'none';
    this.timeout = options.timeout || 30000;
    this.maxRetries = options.maxRetries || 3;
  }

  /**
   * Submit a Proof-of-Computation Summary
   */
  async submitPCS(pcs: PCS): Promise<VerifyResult> {
    // Validate PCS
    this.validatePCS(pcs);

    // Sign PCS if signing is enabled
    if (this.signingAlg !== 'none') {
      this.signPCS(pcs);
    }

    // Send request with retries
    return this.retryRequest(async () => {
      const response = await fetch(`${this.baseURL}/v1/pcs/submit`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'User-Agent': 'fractal-lba-ts-sdk/0.4.0',
          ...(this.tenantID && { 'X-Tenant-Id': this.tenantID })
        },
        body: JSON.stringify(pcs),
        signal: AbortSignal.timeout(this.timeout)
      });

      // Read response body
      const responseBody = await response.text();

      // Handle response status
      switch (response.status) {
        case 200:
        case 202:
          return JSON.parse(responseBody) as VerifyResult;
        case 401:
          throw new SignatureError('Signature verification failed (401)');
        case 429:
          throw new APIError('Rate limit exceeded (429)', 429, responseBody);
        default:
          throw new APIError(
            `API error (${response.status})`,
            response.status,
            responseBody
          );
      }
    });
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<void> {
    const response = await fetch(`${this.baseURL}/health`, {
      method: 'GET',
      signal: AbortSignal.timeout(this.timeout)
    });

    if (response.status !== 200) {
      throw new APIError(
        `Health check failed with status: ${response.status}`,
        response.status
      );
    }
  }

  /**
   * Validate PCS structure
   */
  private validatePCS(pcs: PCS): void {
    if (!pcs.pcs_id) {
      throw new ValidationError('pcs_id is required');
    }
    if (pcs.schema !== 'fractal-lba-kakeya') {
      throw new ValidationError(`Invalid schema: ${pcs.schema}`);
    }
    if (pcs.coh_star < 0 || pcs.coh_star > 1.05) {
      throw new ValidationError(`coh_star out of bounds: ${pcs.coh_star}`);
    }
    if (pcs.r < 0 || pcs.r > 1) {
      throw new ValidationError(`r out of bounds: ${pcs.r}`);
    }
    if (pcs.budget < 0 || pcs.budget > 1) {
      throw new ValidationError(`budget out of bounds: ${pcs.budget}`);
    }
  }

  /**
   * Sign PCS using Phase 1 canonical signing (8-field subset, 9-decimal rounding)
   */
  private signPCS(pcs: PCS): void {
    if (this.signingAlg === 'hmac') {
      if (!this.signingKey) {
        throw new SignatureError('HMAC key not configured');
      }

      // Create signature subset (Phase 1 spec: 8 fields, 9-decimal rounding)
      const subset = {
        budget: round9(pcs.budget),
        coh_star: round9(pcs.coh_star),
        D_hat: round9(pcs.D_hat),
        epoch: pcs.epoch,
        merkle_root: pcs.merkle_root,
        pcs_id: pcs.pcs_id,
        r: round9(pcs.r),
        shard_id: pcs.shard_id
      };

      // Canonical JSON (sorted keys, no spaces)
      const canonicalJSON = JSON.stringify(subset, Object.keys(subset).sort());

      // SHA-256 digest
      const digest = crypto.createHash('sha256').update(canonicalJSON).digest();

      // HMAC-SHA256
      const hmac = crypto.createHmac('sha256', this.signingKey);
      hmac.update(digest);
      const signature = hmac.digest();

      // Base64 encode
      pcs.sig = signature.toString('base64');
    } else {
      throw new SignatureError(`Unsupported signing algorithm: ${this.signingAlg}`);
    }
  }

  /**
   * Retry request with exponential backoff and jitter
   */
  private async retryRequest<T>(
    fn: () => Promise<T>,
    attempt: number = 0
  ): Promise<T> {
    try {
      return await fn();
    } catch (error) {
      // Don't retry validation or signature errors
      if (error instanceof ValidationError || error instanceof SignatureError) {
        throw error;
      }

      // Don't retry if max retries reached
      if (attempt >= this.maxRetries) {
        throw error;
      }

      // Exponential backoff with jitter: base_delay * 2^attempt + jitter
      const baseDelay = 1000; // 1 second
      const maxJitter = 1000; // 1 second
      const delay = baseDelay * Math.pow(2, attempt) + Math.random() * maxJitter;

      // Wait before retrying
      await new Promise(resolve => setTimeout(resolve, delay));

      // Retry
      return this.retryRequest(fn, attempt + 1);
    }
  }
}

/**
 * Round a number to 9 decimal places (Phase 1 spec)
 */
function round9(x: number): number {
  return Math.round(x * 1e9) / 1e9;
}
